pyoomph/ast2ir.py

```diff
@@ -110,33 +207,54 @@ def implicit_conversion(
             var = new_var
         return var
 
+    # Can be called multiple times, that doesn't matter
     def do_args(
-        self, args: List[ast.Expression], target_types: List[Type]
+        self,
+        args: List[ir.LocalVariable],
+        target_types: List[Type],
+        self_var: Optional[ir.LocalVariable],
     ) -> List[ir.LocalVariable]:
-        assert len(args) == len(target_types)
+        if self_var is not None:
+            args = [self_var] + args
+
+        assert len(args) == len(target_types), "wrong number of args"
         return [
-            self.implicit_conversion(self.do_expression(expr), typ)
-            for expr, typ in zip(args, target_types)
+            self.implicit_conversion(var, typ) for var, typ in zip(args, target_types)
         ]
 
-    def do_call(self, call: ast.Call) -> Optional[ir.LocalVariable]:
+    def _get_method_functype(self, the_type: Type, name: str) -> FunctionType:
+        try:
+            return the_type.methods[name]
+        except KeyError:
+            raise RuntimeError(f"{the_type.name} has no method {name}()")
+
+    def do_call(
+        self, call: ast.Call, must_return_value: bool
+    ) -> Optional[ir.LocalVariable]:
         if isinstance(call.func, ast.GetAttribute):
-            self_arg = self.do_expression(call.func.obj)
+            self_var = self.do_expression(call.func.obj)
+            args = [self.do_expression(expr) for expr in call.args]
             try:
-                functype = self_arg.type.methods[call.func.attribute]
+                self._get_rid_of_auto_in_var(self_var, recursive=False)
             except KeyError:
-                raise RuntimeError(
-                    f"{self_arg.type.name} has no method {call.func.attribute}()"
-                )
-            assert self_arg.type == functype.argtypes[0]
-            if functype.returntype is None:
-                result_var = None
+                # Self variable has to remain auto-typed, less type information
+                # available, do_args will be called later
```

Comment:
I wonder if this can be done unconditionally. I think I tried it and it somehow didn't work, but maybe I have changed something so that it would work?

Reply:
Can't do, creates weird errors